[bnb] Minor modifications (#18631)
huggingface/transformers: tests/mixed_int8/README.md

# Testing mixed int8 quantization

![HFxbitsandbytes.png](https://s3.amazonaws.com/moonup/production/uploads/1660567705337-62441d1d9fdefb55a0b7d12c.png)

Hi there, this is a recipe on how to effectively debug `bitsandbytes` integration on HuggingFace `transformers`.

## Library requirements

+ `transformers==4.22.0.dev`

> + `accelerate==0.12.0` 

Reviewer: fixed `==` versions should be avoided by all means as it wrecks havoc with virtual envs, which leads to conflicting requirements

> + `accelerate==0.12.0` 

Author: Totally makes sense! Thanks for that

+ `0.31.8>=bitsandbytes>=0.31.5`.
## Hardware requirements

I am using a setup of 2 GPUs that are NVIDIA-Tesla T4 15GB - `younes-testing-multi-gpu` on GCP. To run successfully `bitsandbytes` you would need a 8-bit core tensor supported GPU. Note that Ampere architectures should be supported. Here is an exhaustive list of the supported GPU types at the time of writing:

- RTX 20s & RTX 30s
- A40-A100

> - T4 + (e.g. Google Colab GPUs) 

Author: I proposed a change in 632a5a73c8c950b9b925454dd7f9baa435f7841e !

> - T4 + (e.g. Google Colab GPUs) 

Author: Great thanks ! I proposed a fix in 82a9c8edf0292875a37debb40724b39d1f3d3608 & e9502267a270006a9cfad1acc68cfee290880c75

## Virutal envs

```conda create --name int8-testing python==3.8```
```pip install bitsandbytes```
```pip install accelerate```

> ```pip install git+https://github.com/huggingface/transformers.git```

Author: ```suggestion
```pip install git+https://github.com/huggingface/transformers.git```
if `transformers>=4.22.0` is not released yet, otherwise
```pip install transformers```
```

## Trobleshooting

A list of common errors:

### Torch does not correctly do the operations on GPU

First check that:

```py
import torch

vec = torch.randn(1, 2, 3).to(0)
```

Works without any error. If not install torch using `conda` like:

```conda create --name int8-testing python==3.8```
```pip install bitsandbytes```
```conda install pytorch torchvision torchaudio -c pytorch```
```pip install git+https://github.com/huggingface/transformers.git```

> ```pip install accelerate```

Author: Thanks! Resolved in 0b17a890cf7828e83873128cdb7a19662bd3a724

# Testing mixed int8 quantization

![HFxbitsandbytes.png](https://s3.amazonaws.com/moonup/production/uploads/1660567705337-62441d1d9fdefb55a0b7d12c.png)

Hi there, this is a recipe on how to effectively debug `bitsandbytes` integration on HuggingFace `transformers`.

## Library requirements

+ `transformers>=4.22.0`
+ `accelerate>=0.12.0` 
+ `0.31.8>=bitsandbytes>=0.31.5`.
## Hardware requirements

I am using a setup of 2 GPUs that are NVIDIA-Tesla T4 15GB - `younes-testing-multi-gpu` on GCP. To run successfully `bitsandbytes` you would need a 8-bit core tensor supported GPU. Note that Ampere architectures should be supported. Here is an exhaustive list of the supported GPU types at the time of this writing:

- RTX 20s & RTX 30s
- A40-A100
- T4 + (e.g. Google Colab GPUs) 

## Virutal envs

```conda create --name int8-testing python==3.8```

> ```pip install bitsandbytes```
> ```pip install accelerate```
> ```pip install git+https://github.com/huggingface/transformers.git```

Author: ```suggestion
```pip install bitsandbytes>=0.31.5```
```pip install accelerate>=0.12.0```
```pip install git+https://github.com/huggingface/transformers.git```
```

pip install git+https://github.com/huggingface/transformers.git
```

> ## Trobleshooting

Author: ```suggestion
## Troubleshooting
```

# Testing mixed int8 quantization

![HFxbitsandbytes.png](https://s3.amazonaws.com/moonup/production/uploads/1660567705337-62441d1d9fdefb55a0b7d12c.png)

Hi there, this is a recipe on how to effectively debug `bitsandbytes` integration on HuggingFace `transformers`.

## Library requirements

+ `transformers>=4.22.0`
+ `accelerate>=0.12.0` 

> + `0.31.8>=bitsandbytes>=0.31.5`.

Reviewer 2: Is `0.31.8` necessary? Or it is just the version that has been tested without problem?

Author: Actually we know that these versions work and the latest version is `0.31.8` , so yes it's the versions we know it has been tested without problem

Author: Actually there is no reason, proposed an update below
```suggestion
+ `bitsandbytes>=0.31.5`.
```

## Hardware requirements

> I am using a setup of 2 GPUs that are NVIDIA-Tesla T4 15GB - `younes-testing-multi-gpu` on GCP. To run successfully `bitsandbytes` you would need a 8-bit core tensor supported GPU. Note that Ampere architectures should be supported. Here is an exhaustive list of the supported GPU types at the time of this writing:

Reviewer 2: I would avoid using the first person pronouns like `I am ...` in writing.

It would be more natural to be `The following instructions are tested with 2 NVIDIA-Tesla T4 GPUs`.

Author: ```suggestion
The following instructions are tested with 2 NVIDIA-Tesla T4 GPUs. To run successfully `bitsandbytes` you would need a 8-bit core tensor supported GPU. Note that Ampere architectures should be supported. Here is an exhaustive list of the supported GPU types at the time of this writing:
```

Author: Thanks for the suggestion!

Reviewer 2: It's also unnecessary to add `younes-testing-multi-gpu` on GCP`.
(This doesn't give any useful information to the readers)
(It's also best not to reveal information regarding internal infrastructure)

Author: Yep! Sorry for that 🙏

Author: Should fixed in 8e70e5f70384bead7009ad628a8e43f8ceb47d2e

- RTX 20s & RTX 30s
- A40-A100
- T4 + (e.g. Google Colab GPUs) 

## Virutal envs

```conda create --name int8-testing python==3.8```
```pip install bitsandbytes>=0.31.5```
```pip install accelerate>=0.12.0```
```pip install git+https://github.com/huggingface/transformers.git```
if `transformers>=4.22.0` is not released yet, otherwise
```pip install transformers```

## Trobleshooting

A list of common errors:

### Torch does not correctly do the operations on GPU

First check that:

```py
import torch

vec = torch.randn(1, 2, 3).to(0)
```

> Works without any error. If not install torch using `conda` like:

Author: ```suggestion
Works without any error. If not, install torch using `conda` like:
```

```conda create --name int8-testing python==3.8```
```pip install bitsandbytes```
```conda install pytorch torchvision torchaudio -c pytorch```
```pip install git+https://github.com/huggingface/transformers.git```
```pip install accelerate```

> And the snippet above should work

Reviewer 2: ```suggestion
And the snippet above should work.
```

> ### ` `bitsandbytes` operations are not supported under CPU!`

Reviewer 2: The usage of **`** here makes the rendered version (GitHub page) looks strange.

see https://github.com/younesbelkada/transformers/blob/bnb-fix-small-details/tests/mixed_int8/README.md

Author: Thanks for spotting this!
Should be fixed in c8d628195bd9fca4ae25031205d220aa2248ab70

This happens when some Linear weights are set to the CPU when using `accelerate`. Please check carefully `model.hf_device_map` and make sure that there is no `Linear` module that is assigned to CPU. It is fine to have the last module (usually the Lm_head) set on CPU.

### `To use the type as a Parameter, please correct the detach() semantics defined by __torch_dispatch__() implementation.`

Use the latest version of `accelerate` with a command such as: `pip install --force accelerate` and the problem should be solved.

### `Parameter has no attribue .CB` 

> Same comment as above

Reviewer 2: ```suggestion
Same comment as above.
```

### `RuntimeError: CUDA error: an illegal memory access was encountered ... consider passing CUDA_LAUNCH_BLOCKING=1`

> Run your script by pre-pending `CUDA_LAUNCH_BLOCKING=1` and you should observe an error as below:

Reviewer 2: As the next line is with `###`, here it's better to be `... you should observer an error described in the next section.`

Author: Totally agree yes! fixed in c8d628195bd9fca4ae25031205d220aa2248ab70

### `CUDA illegal memory error: an illegal memory access at line...`:

Check the CUDA verisons with:
```
nvcc --version
```

> And confirm it is the same version than the one detected by `bitsandbytes`. If not, run:

Reviewer 2: ```suggestion
And confirm it is the same version as the one detected by `bitsandbytes`. If not, run:
```

Reviewer 2: Or just `Confirm it is ...`

```
ls -l $CONDA_PREFIX/lib/libcudart.so
```
or 
```
ls -l $LD_LIBRARY_PATH
```

> And check if `libcudart.so` has a correct symlink that is set. Sometimes `nvcc` detects the correct CUDA version but `bitsandbytes` doesn't. You have to make sure that the symlink that is set for the file `libcudart.so` is redirected to the correct CUDA file. 

Reviewer 2: you can use `and`, or just `Check if ...`

Author: Perfect! Fixed it in c8d628195bd9fca4ae25031205d220aa2248ab70

Here is an example of a badly configured CUDA installation:

`nvcc --version` gives:

![Screenshot 2022-08-15 at 15.12.23.png](https://s3.amazonaws.com/moonup/production/uploads/1660569220888-62441d1d9fdefb55a0b7d12c.png)

> Which means that the detected CUDA version is 11.3 but `bitsandbytes` outputs:

Reviewer 2: ```suggestion
which means that the detected CUDA version is 11.3 but `bitsandbytes` outputs:
```

![image.png](https://s3.amazonaws.com/moonup/production/uploads/1660569284243-62441d1d9fdefb55a0b7d12c.png)

Therefore check:

```
ls -l /opt/conda/envs/py37/lib/libcudart.so
```

> And you can see that:

Reviewer 2: ```suggestion
and you can see that:
```

![Screenshot 2022-08-15 at 15.12.33.png](https://s3.amazonaws.com/moonup/production/uploads/1660569176504-62441d1d9fdefb55a0b7d12c.png)

> If you see that the file is linked to the wrong CUDA version (here 10.2), find the correct location for `libcudart.so` (`find --name ...`) and replace the environment variable `LD_LIBRARY_PATH` with the one containing the correct `libcudart.so` file.

Reviewer 2: Might be a good idea to provide the full command regarding `find --name`

Author: Yes! Added the full command in c8d628195bd9fca4ae25031205d220aa2248ab70

### If `bitsandbytes` installation breaks everything:

It happened in a previous version that after installing `bitsandbytes` and running this script:

```py
import bitsandbytes as bnb
```

> You get an error:

Reviewer 2: ```suggestion
we got an error:
```

pip install bitsandbytes>=0.31.5
pip install accelerate>=0.12.0
pip install git+https://github.com/huggingface/transformers.git
```
if `transformers>=4.23.0` is not released yet, otherwise
```pip install transformers```

## Trobleshooting

A list of common errors:

### Torch does not correctly do the operations on GPU

First check that:

```py
import torch

vec = torch.randn(1, 2, 3).to(0)
```

Works without any error. If not, install torch using `conda` like:

```bash
conda create --name int8-testing python==3.8
pip install bitsandbytes
conda install pytorch torchvision torchaudio -c pytorch
pip install git+https://github.com/huggingface/transformers.git
pip install accelerate
```

and the snippet above should work.

### ` bitsandbytes operations are not supported under CPU!`

This happens when some Linear weights are set to the CPU when using `accelerate`. Please check carefully `model.hf_device_map` and make sure that there is no `Linear` module that is assigned to CPU. It is fine to have the last module (usually the Lm_head) set on CPU.

### `To use the type as a Parameter, please correct the detach() semantics defined by __torch_dispatch__() implementation.`

> Use the latest version of `accelerate` with a command such as: `pip install --force accelerate` and the problem should be solved.

Author: ```suggestion
Use the latest version of `accelerate` with a command such as: `pip install -U accelerate` and the problem should be solved.
```

> Use the latest version of `accelerate` with a command such as: `pip install --force accelerate` and the problem should be solved.

Author: You are right here, updated above!

```
or 
```
ls -l $LD_LIBRARY_PATH
```
Check if `libcudart.so` has a correct symlink that is set. Sometimes `nvcc` detects the correct CUDA version but `bitsandbytes` doesn't. You have to make sure that the symlink that is set for the file `libcudart.so` is redirected to the correct CUDA file. 

Here is an example of a badly configured CUDA installation:

`nvcc --version` gives:

![Screenshot 2022-08-15 at 15.12.23.png](https://s3.amazonaws.com/moonup/production/uploads/1660569220888-62441d1d9fdefb55a0b7d12c.png)

which means that the detected CUDA version is 11.3 but `bitsandbytes` outputs:

![image.png](https://s3.amazonaws.com/moonup/production/uploads/1660569284243-62441d1d9fdefb55a0b7d12c.png)

Therefore check:

```
ls -l /opt/conda/envs/py37/lib/libcudart.so
```

and you can see

![Screenshot 2022-08-15 at 15.12.33.png](https://s3.amazonaws.com/moonup/production/uploads/1660569176504-62441d1d9fdefb55a0b7d12c.png)

If you see that the file is linked to the wrong CUDA version (here 10.2), find the correct location for `libcudart.so` (`find --name libcudart.so`) and replace the environment variable `LD_LIBRARY_PATH` with the one containing the correct `libcudart.so` file.

> ### If `bitsandbytes` installation breaks everything:

Author: Thanks for the feedback! As you said below, I think that setting the correct version of bitsandbytes should be sufficient

> ### If `bitsandbytes` installation breaks everything:

Author: Removed in e16a56a67d8da84e111b7017821d257fc941d5f6

It happened in a previous version that after installing `bitsandbytes` and running this script:

```py
import bitsandbytes as bnb
```

we got an error:

```
major, minor, revision = ...
Too many values to unpack...
```

> Re-install `bitsandbytes==0.31.8` or `bitsandbytes==0.31.5` as everything worked fine on our Docker image with those versions. In the worst case remove the [line that installs bitsandbytes on the Dockerfile](https://github.com/huggingface/transformers/blob/d6eeb871706db0d64ab9ffd79f9545d95286b536/docker/transformers-all-latest-gpu/Dockerfile#L49)

Author: I have the same comments as above!
